### DISCUSSION TITLE
.NET 8 Blazor Web App Migration (Continued...)

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3798] [General]
The previous thread (https://github.com/oqtane/oqtane.framework/discussions/3698) is getting too long, so starting a new one.

Now that most of the UI use cases are working, it is time to focus on the runtime workload characteristics...

When running on Interactive rendering, the App server component is executed once for every unique browser session. It makes a number of database calls to retrieve information so that it can populate the initial page markup. It then initiates the component rendering process which intantiates the SiteRouter. The SiteRouter makes a number of Service API calls (using HttpClient) to populate the PageState. On each subsequent user interaction the SiteRouter utilizes the existing PageState so that it does not have to reload information.

When running on Static rendering currently, the App server component is executed for every request (ie. every user interaction). It makes a number of database calls to retrieve information so that it can populate the initial page markup. It then initiates the component rendering process which intantiates the SiteRouter. The SiteRouter makes a number of Service API calls (using HttpClient) to populate the PageState. On each subsequent request (ie. user interaction) the SiteRouter has to reload the PageState.

So based on above, it is clear that when running on Static rendering the framework currently has a much higher workload than it does on Interactive rendering. In fact, on each request it is loading the same information multiple times (ie. in both the App and SiteRouter components). In addition, the SiteRouter is relying on Service API calls (using HttpClient) which is not necessary in Static rendering, as this workload is running on the server. 

In order to address the runtime characteristics above, I am going to isolate the workload which is currently in SiteRouter and move it to App. App will construct the PageState and will pass it as a component parameter to Routes which will then pass it to SiteRouter. SiteRouter will then use the PageState if it is provided (which will not always be the case as if you are running on .NET MAUI then App is never executed). This will eliminate the "double workload" which currently exists for Static rendering. It will also benefit Interactive rendering (to a lesser extent as App and SiteRouter are not being executed frequently in interactive scenarios - however if you are running on Globally Interactive public site which contains a lot of content pages then this improvement will be beneficial).

I will also be implementing the approach which was demonstrated in the Oqtane SSR POC which utilizes dependency injection to vary the implementation of Service APIs based on where they are executing (ie. server or client). This will allow server-based workloads to avoid the HttpClient overhead... and client-based workloads to continue to function as they do today.

Again, I am giving a heads up about these changes because they are invasive to the framework and may result in some instability in the Dev branch for a period of time.

**Comment by sbwalker** [comment]
The changes above are mostly complete - the SiteService has been refactored to have client and server implementations, and the App component now uses the SiteService to retrieve information using the server implementation and passes the PageState to the SiteRouter - which eliminates a number of data access calls on a full page request when running on Static or Interactive render mode. 

One of the challenges I am working on currently is related to the Head component. In Interactive rendering this component is responsible for injecting the favicon, page title, as well as any arbitrary head content (ie. meta, link references, etc...) based on the configuration in your site. This component is also what allows modules to dynamically set the page title, etc... It accomplishes this using the INotifyPropertyChanged concept which is part of Blazor. 

Unfortunately, this approach does not seem to work in Static rendering. There are a number of issues logged in the aspnetcore repo related to HeadOutlet - which suffers from a similar problem. Basically the problem is that although the Head component receives the INotifyPropertyChanged events, it does not render the markup in the UI. This seems to be related to "mixed mode" scenarios ie. the app loads in Static render mode by default and has a few components which are Interactive. In this configuration the DOM differencing algorithm gets confused and actually removes elements from the DOM which were added during static/pre rendering (see https://github.com/dotnet/aspnetcore/issues/50268). Again this is part of the Blazor "black box" which makes it difficult to diagnose and find workarounds.

This is a fairly major problem as it means that  favicon, page title, as well as any arbitrary head content (ie. meta, link references, etc...) are not being injected into the page - which are some of the most critical elements you want to include in a static rendering scenario (for SEO). I initially tried implementing some logic in the App component to include elements in the page head and although it would work for configuration stored in the site, it would not support the dynamic ability for a module to set the title, etc... So more research is required. This link offers some advanced render mode concepts but I have not been able to get any of them to work for this specific scenario - https://learn.microsoft.com/en-us/aspnet/core/blazor/components/render-modes?view=aspnetcore-8.0#fine-control-of-render-modes

Another challenge is related to module and theme scripts and stylesheets. In Interactive rendering, scripts and stylesheets are handled by JS Interop in OnAfterRender() - which is implemented in ModuleBase, ThemeBase, etc... In Static rendering you can't use JS Interop and the OnAfterRender() event is not even called so a different approach is needed. Rather than having 2 totally different implementations I am hoping I can find a happy medium. I am currently experimenting with calling the JavaScript LoadJS methods in the traditional way when running on Static rendering (rather than using JS Interop). This would provide consistency on how scripts are injected into the DOM - they would just be triggered in different ways by the framework based on the render mode.

**Comment by sbwalker** [comment]
I have not been able to make any progress on the INotifyPropertyChanged issue with Static rendering. If you want to experiment with the problem yourself open the Head.razor file and put some breakpoints on the various methods. 

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/c6f02ffa-3caf-4e6c-bf42-4d59f3798a48)

Also put a breakpoint in ThemeBuilder,razor where it sets the page title:

```
        // set page title
        if (!string.IsNullOrEmpty(PageState.Page.Title))
        {
            SiteState.Properties.PageTitle = PageState.Page.Title;
        }
        else
        {
            SiteState.Properties.PageTitle = PageState.Page.Name + " - " + PageState.Site.Name;
        }
```

When you run the app you will see the OnInitialized() method in Head.razor get called first, registering the PropertyChanged event. Then you will see the ThemeBuilder.razor logic called where it sets the PageTitle. Then the Head.razor PropertyChanged event will fire and you can see the _title property get set with a value. This should be rendered in App.razor where the Head.razor is declared:

```
    <!DOCTYPE html>
    <html lang="@_language">
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <base href="/" />
        <script src="js/app.js"></script>
        <script src="js/loadjs.min.js"></script>
        <link rel="stylesheet" href="css/app.css" />
        @if (!string.IsNullOrEmpty(_PWAScript))
        {
            <link id="app-manifest" rel="manifest" />
        }
        @((MarkupString)_styleSheets)
        <link id="app-stylesheet-page" />
        <link id="app-stylesheet-module" />
        @if (_renderMode == RenderModes.Static)
        {
            <Head />
        }
        else
        {
            <Head @rendermode="InteractiveRenderMode.GetInteractiveRenderMode(_runtime, _prerender)" />
        }
        @((MarkupString)_headResources)
    </head>
```

However the markup is never rendered into the page output:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/1b6355f1-2d2e-4a81-baac-74fb4103af93)

Note that if you set the Site to use Interactive render mode, this approach works fine and includes the page title:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/1bb67da6-f804-49b3-888c-bb4eb40a2ba5)

The strange thing is that is that this same problem exists in the Oqtane SSR POC... however I also created another Blazor Web App with per component rendering and added a service with implements INotifyPropertyChanged to it... and it renders the content to the UI as I would expect. So I am trying to determine what the difference is which is causing the problem.

NotifyProperty.cs
```
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace BlazorAppWithAuth.Models
{
    public class NotifyProperty : INotifyPropertyChanged
    {
        public event PropertyChangedEventHandler? PropertyChanged;

        private string _value = "";

        public string Value
        {
            get => _value;
            set
            {
                if (_value != value)
                {
                    _value = value;
                    OnPropertyChanged();
                }
            }
        }

        protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = default)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}
```
Program.cs
```
            builder.Services.AddScoped<NotifyProperty>();
```

Sender.razor
```
@using BlazorAppWithAuth.Models
@inject NotifyProperty NotifyProperty

@code {
    [Parameter]
    public string Value { get; set; } = "";

    protected override void OnParametersSet()
    {
        NotifyProperty.Value = $"<title>{Value}</title>";
    }
}
```
Receiver.razor
```
@using System.ComponentModel
@using BlazorAppWithAuth.Models
@inject NotifyProperty NotifyProperty

@if (!string.IsNullOrEmpty(_value))
{
    @((MarkupString)_value)
}

@code {
    private string _value = "";

    protected override void OnInitialized()
    {
        ((INotifyPropertyChanged)NotifyProperty).PropertyChanged += PropertyChanged;
    }

    private void PropertyChanged(object? sender, PropertyChangedEventArgs? e)
    {
        if (NotifyProperty.Value != _value)
        {
            _value = NotifyProperty.Value;
            StateHasChanged();
        }
    }

    public void Dispose()
    {
        ((INotifyPropertyChanged)NotifyProperty).PropertyChanged -= PropertyChanged;
    }
}
```
App.razor
```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <base href="/" />
    <link rel="stylesheet" href="bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" href="app.css" />
    <link rel="stylesheet" href="BlazorAppWithAuth.styles.css" />
    <link rel="icon" type="image/png" href="favicon.png" />
    <Receiver />
</head>
<body>
    <Routes />
    <script src="_framework/blazor.web.js"></script>
</body>
</html>
```
Home.razor
```
@page "/"

<h1>Hello, world!</h1>

Welcome to your new app.

<Sender Value="HOME" />
```
Counter.razor
```
@page "/counter"
@rendermode InteractiveServer

<h1>Counter</h1>

<p role="status">Current count: @currentCount</p>

<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

<Sender Value="COUNTER" />

@code {
    private int currentCount = 0;

    private void IncrementCount()
    {
        currentCount++;
    }
}
```
Results in:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/7465d770-d71c-43c5-9d0f-d9ba1d558fe3)

**Reply by sbwalker** [reply]
Another interesting observation... there are other aspects of Oqtane which rely on the INotifyPropertyChanged capability of SiteState. For example, it is possible for a module to programmatically set its own module title using the SetModuleTitle() method - which internally uses SiteState and INotifyPropertyChanged. If I add this method to the Html/Text module's Index.razor:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/c954cd7b-a9b5-4b7e-bf3a-bde367747d7d)

It produces the following output:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/c977c694-9919-4fe4-882d-31005300ceda)

So it works fine on Static rendering in this scenario. The only scenario where it does not seem to work is in the Head.razor component.

**Reply by sbwalker** [reply]
It appears that as long as the INotifyPropertyChanged event is processed by a component within the same render tree of nested components it will render the content... however since Head is a sibling component to Routes the INotifyPropertyChanged event is triggered but the component does not render the content. I wonder if this is because of some process (not render mode) boundary.

**Reply by thabaum** [reply]
@sbwalker so to be clear, correct me if I am wrong, our current objective is to find a way to get the `_title` variable to be passed to the `<title></title>`?
I notice the blinking in Static Server that does not exist as much or at all in Interactive Server render mode.  Like things are loading twice when visiting the interactive admin component pages.  Page Management I think along with System Info seem to be the smoothest or more expected behavior.

I want to ask a couple other follow up questions.

If I recall I may have read something in the ASP.NETCORE issues after you pointed us to yours to which flowed into many other issues similar, and a fix coming in .NET 9 if I am not mistaken.  So might be harder to fix something that has a crack in the foundation.  If true, maybe we need to look closer at that solution Microsoft is using for clues on a fix and come up with our own version as a spare tire until .NET 9?

I have been trying to recall if there was a mention of things being overwritten with blank information the second time rendering. I can try to get those links to you as it relates to everyone losing head values if I am not mistaken.  And as you go down that rabbit hole you eventually find one that points to an issue as resolved in .NET 9.  So I have been investigating this area as well reviewing [current preview commits.](https://github.com/dotnet/aspnetcore/commits/v9.0.0-preview.1.24081.5/)

Here are a couple other interesting articles that may relate a little and give clues on how we can handle this current issue hopefully:
This pull request thread has some interesting reads on data permanent/enhance elements:
- [[Blazor] APIs to preserve content between enhanced navigations](https://github.com/dotnet/aspnetcore/pull/50437)

This issue has links at the bottom of the thread to a few other relating issues that are good to be fully aware of I believe: 
- [Proper discovery of SSR pages/endpoints for Blazor](https://github.com/dotnet/aspnetcore/issues/46980#top)

We cannot pass an id and such in the head, which is why I know I read another issue relating to specifically the head such as a title that took a few links in conversations to find.

I have verified testing the issue exists as described and will work with both Oqtane and the POC to see if I can find any further clues to help us find a solution.

**Reply by sbwalker** [reply]
@thabaum the current objective is to find a way for components to be able to communicate state to the Head root component when running on Static rendering. Interactive rendering uses INotifyPropertyChanged for this purpose and I was hoping we could use the same approach for Static rendering. Actually there is no other solution which I am aware of which could satisfy all of the requirements we have in Oqtane:

- setting page title dynamically from components
- adding meta tags dynamically from components
-  adding link references (CSS) dynamically from components
- adding scripts dynamically from components

This is core functionality and must have a solution in order for Static rendering to be viable.

I will explore the links you provided to see if there are any clues. The fact that I can get INotifyPropertyChanged to work as expected in a vanilla Blazor Web project would lead me to believe that there must be a solution for Oqtane. And I agree that waiting for .NET 9 is not an option.

**Reply by sbwalker** [reply]
In regards to the links above:

data-permanent is intended to preserve UI markup across page navigations. The Oqtane SSR POC has an example of this with the Search box in the main layout - it preserves the search criteria which has been entered even when a user clicks various navigation options. This does not relate to the INotifyPropertyChanged problem as there is no page navigation in this scenario - it is happening on the initial page construction.

The item related to endpoint discovery is also not related to Oqtane as Oqtane does not use the standard Blazor router or page components (ie. ```@page```). It uses a fallback endpoint and then uses its own custom SiteRouter.

**Reply by thabaum** [reply]
Since this area is static and not interactive we may need to provide it with a JavaScript to get by on?

> 
> Some apps depend on JS to perform initialization tasks that are specific to each page. When using Blazor's enhanced navigation feature, which allows the user to avoid reloading the entire page, page-specific JS may not be executed again as expected each time an enhanced page navigation occurs.
> 
> To avoid this problem, we don't recommended relying on page-specific <script> elements placed outside of the layout file applied to the component. Instead, scripts should register an [afterWebStarted JS initializer](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/startup?view=aspnetcore-8.0#javascript-initializers) to perform initialization logic and use an event listener (blazor.addEventListener("enhancedload", callback)) to listen for page updates caused by enhanced navigation.

[ASP.NET Core Blazor JavaScript with static server-side rendering (static SSR)](https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/static-server-rendering?view=aspnetcore-8.0)

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#enhanced-navigation-and-form-handling

Both may have some clues as well in resolving any future issues for developers trying to make things work in static SSR.  I have also read on how we may use 2 layouts but not sure that should be the case.  I would imagine one with the JavaScript file hard coded and another without depending on global interactivity or not?  I would think we would just add the script file or not depending on the mode running for the tenant?

**Reply by sbwalker** [reply]
Another observation... if I comment out the logic which sets the page title in ThemeBuilder.razor and add the following logic to Routes.razor:

```
    protected override async Task OnParametersSetAsync()
    {
        SiteState.Properties.PageTitle = DateTime.Now.ToString();
```

The page title element is rendered:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/4a7e91f1-9cb9-4315-bd98-4f82df5b6329)

So this proves that when using Static rendering, state can be communicated between components using INotifyPropertyChanged. However, it does not work when invoked within some components... so now to try and determine the reason.

**Reply by thabaum** [reply]
I was noticing how SiteState.Properties.PageTitle is null, so it would never put the title in the head section.  If I placed it in the onintitialized override the element would parse without the page title.  I put break points on that property and seen it never gets set (null).

**Reply by thabaum** [reply]
@sbwalker hate to say it but could this fix it... adding `@attribute [StreamRendering]` to the Head.razor made it work.  Can you verify?

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/b73ca197-9955-49a4-93cd-98745db46867)

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/0e9842cc-db5d-47e8-84fe-2fe6495aefa9)

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/d6da820c-603e-4877-8c1a-8d72c1922b60)

Save changes to new home page title

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/ed784c91-1599-4aac-a9bb-4d301bb2697e)

**Reply by sbwalker** [reply]
@thabaum yes, that does appear to work. I tested it in static and interactive rendering and it works for both. I also tried to move the [StreamRendering] attribute to the App.razor (since both Head and Routes seem to need it to function properly) but that did not work at all - the browser just showed an empty page. I would really like to better understand what [StreamRendering] is actually doing to resolve some of these strange issues. Either way, this unblocks a major issue for static rendering.

**Reply by thabaum** [reply]
@sbwalker StreamRendering sends updates to the browser regularly to some regard.  It is suggested by Microsoft that component library authors all use stream rendering as it will work best with all global render modes (static and interactive modes).

The issue I am stuck with now is more relating to the CSS or JS not loading for the modules so the theme css styles for those components work.  The buttons on the navigation work as expected now.   But the login form uses 100% so that surely says this part is still not interactively working properly.  I have been testing different areas to place this stream rendering and try to limit it to where it is needed only so it is not called too often.  

As you say I would also like to get to know this stream rendering feature better for example how often it can be called, what are the trade offs if any.

There is also the relating prerendering boolean `@attribute [StreamRendering(true)]` which is true by default, but can be set to false.

**Reply by sbwalker** [reply]
The module CSS and JS is currently not being loaded on static rendering. This is because when running on interactive rendering it is being injected dynamically using JS Interop methods in OnAfterRender. This method is not called on static rendering. Now that INotifyPropertyChanged is working I can inject the module CSS and JS using the Head component. This is one of the reasons why it was so important to resolve the issues with  INotifyPropertyChanged.

**Reply by sbwalker** [reply]
@thabaum PR #3843 now injects the module JS and CSS (using the Head component and INotifyPropertyChanged).

**Reply by thabaum** [reply]
@sbwalker seems to work rendering the login page component css styles, however I have an issue accessing the control panel from the icon after logging in.  
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/9268c1c0-1c64-4d9d-92bb-fc3a28b52b10)
You had fixed this once before, as now the sections are not interactive either sending you back to home page when clicked on to expand in Site Settings as well.
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/14200859-e7fb-47b4-abf7-ae7c9aa7febe)

**Reply by sbwalker** [reply]
@thabaum yes, I am working on that issue now

**Reply by sbwalker** [reply]
@thabaum PR #3847 will unblock you... but it will require more investigation to determine if it has any side effects in interactive rendering

**Comment by sbwalker** [comment]
@thabaum in regards to your earlier comment "I notice the blinking in Static Server that does not exist as much or at all in Interactive Server render mode. Like things are loading twice when visiting the interactive admin component pages. Page Management I think along with System Info seem to be the smoothest or more expected behavior." - I do have some theories about this which I will need to explore tomorrow. 

Basically in interactive rendering, Oqtane is relying on custom JavaScript to efficiently manage the injection of stylesheets... there was a long discussion about this last year when we upgraded to .NET 7 because we tried to migrate to using the new Head component approach but it did not provide a good user experience we so rolled back to the custom method. In static rendering I think something similar is happening where the stylesheet references are being removed and then re-added on every page load - which is resulting in the behavior you are seeing in the browser.

**Reply by thabaum** [reply]
@sbwalker I will let you drive things ahead further while investigating some of my other issues reported like #3839.  I will also pay attention to how things relate to the style sheets loading generating a flicker.  I do not notice this with interactive server everthing works awesome. Thank you for the clarification and nice work! 🚀

**Reply by sbwalker** [reply]
@thabaum PR #3850 should address the flash on static rendering. Basically I was able to validate that the injection of stylesheets needs to occur in the App component. This is because the Head component is rendered later in the process (after the App component in finished its initial render) which results in the initial set of stylesheets not being included in the page content sent to the browser (as a result of StreamRendering) and then being populated later - which produces a flash in the browser. Moving this logic to the App component means that the stylesheets are included in the page content sent to the browser on the initial render. 

Interactive rendering also required stylesheets to be injected in the App component for a similar reason... however it never handled module stylesheets previously (because the data was not available in the App component). With the recent introduction of the SiteService it means that the data is available in the App component so module stylesheets are now included - which benefits both static and interactive scenarios.

**Reply by thabaum** [reply]
@sbwalker that is awesome you were able to move those styles loading that far up the pipeline so to speak.  I was curious if this was possible for Oqtane Framework.  It would be nice to remove the flash.  I am about to download the latest copy and take it for a test run.  

Looks like we have an issue that is very odd #3838 maybe you should review as it maybe considered a Blazor Web App migration issue.  @zyhfish verified the issue and has included extra information along with clues found after installation simply visit Scheduled Jobs which creates a flashing loop of the message I believe a fix was provided to work around, but besides breaking the control panel interactivity the issue bugs the installation of the default jobs.  It happens in both interactivity modes.

Moving ahead each day, rock and roll 🎸

**Reply by thabaum** [reply]
@sbwalker After testing, this works a lot better as the theme is solid, only the UI loads twice on Site Settings which maybe needs a `data-permanent` or `data-enhanced` attribute potentially on a `<div>` or `<form>` element possibly.  I also noticed the installation page loading twice still while working this area.

The double loading is now limited to just the loading of the components.  The theme seems to be have lost this strange behavior.  In fact great, amazing!  I love it.  Barely a blink if any and for static ssr that is great.

For the admin and profile pages it feels like it loads the component and then it disappears and loads it again.  This is supposed to be resolved with those form attributes.  So this form should get loaded, and stay loaded even if reloaded.  So it looks like you still have some things to review in this area for improvement if possible.  Can we do the same thing for loading these page components?
___
After reviewing the newly introduced `SiteService` I notice it still uses some http to get `RoleNames` and `PermissionNames` and `User` for example.  Will these server services be next to produce?

I was wondering how things would be handled for naming if in the same namespace and see you decided on `ServerSiteService` which makes sense and aligns with what I was thinking as one of the options.  I wonder if we should name the file the same so we can easily locate it on a file search although looking in the server project works and is obvious?  

I also am wondering about the interface.  Would we expect to also see introduced another interface for each server-side service that does not use the controllers in the end?  I maybe viewing this architecture a bit wrong so curious how my vision aligns with the end objectives currently.

In other words in the long haul, Oqtane will need its own `IServerSiteService` interface located in the `Oqtane.Server\Services\Interfaces` folder to be created to communicate directly only needing to reference `ServerSiteService` and bypassing the http requests to the controller, only handling the requests using the repository directly or other framework functionality.
___
## 🚀 
I tested the branch latest commit, and wow what a difference!  I no longer see the login resizing.  This is smoothing out the road.  Since things seem to get loaded on the first instance of the page, I am wondering why some of these forms render twice, if by chance the same approaches can be applied here in a sense?  This really is a much needed overhaul and paying dividends right now from what I am witnessing.  Keep up the great work and thanks again!

**Reply by sbwalker** [reply]
@thabaum #3838 is fixed - very simple solution - AddModuleMessage() was being called in OnParametersSet which automatically triggers a UI refresh... which will call OnParametersSet again... infinite loop. Moving the logic to OnInitialized resolves the issue.

**Reply by sbwalker** [reply]
@thabaum in regards to your other questions... the "double loading" you are seeing is for any interactive component which is being rendered in a static page. These components render once statically and then render a second time interactively. This is how a mixed render mode is expected to behave. The data-permanent attribute is only for retaining content when navigating to a new page - in this case you are not navigating. The data-enhance is only for forms which are statically rendered - and in this case the modules are rendered interactively. I think the double loading of the UI is caused by the stream rendering... as it is essentially rendering the UI multiple times within the same request - but if we remove this attribute, the app does not function at all. So I am not sure it can be optimized any further.

In regards to the SiteService, I would like to follow the existing naming convention. Basically there are 2 implementations of the same interface - one which is used in client scenarios and one which is used in server scenarios - all controlled by dependency injection at runtime. The SiteService on the client uses HttpClient to call the SiteController on the server which calls the SiteService on the server where the actual implementation resides. The SiteService on the server (Oqtane.Server\Services\SiteService.cs) does not utilize HttpClient - it only calls repositories.  

The SiteService is responsible for the bulk of the workload during startup and page rendering in Oqtane. This is why it made sense to abstract it so that it could be called by both the client and server. In the future it might make sense to also create client and server implementations of the other Service interfaces which exist in the Client project currently. This would not really benefit the core framework UI processing but it would provide benefits for specific modules, and it would create a consistent API for core services regardless of whether you are developing static modules or interactive modules. However there is a lot of work involved to do this  - it is not simply a cut and paste exercise - each method requires some modification to adapt to the changes which means there is a high chance of regression issues surfacing. So I would prefer to delay this effort to a future release and focus on stabilizing the current code base.

**Reply by thabaum** [reply]
@sbwalker I may have figured it out... put `@attribute [StreamRendering(false)]` on the `Site\Index.razor`

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/20b99d9c-2100-46d9-9618-f996174538a8)

Only should need this on components that act like this and may need this added to stop the prerendering double loading behavior.   This placed in the page component itself took away the double load behavior.  Putting this further up the pipeline broke other module components first noticeable in the HTML component.

I will let you play with this new found behavior and how it should be applied where once you can get a feel for it.  Another thing to think about while stream rendering that looks like can play a big part in how a component behaves.

It's like the old saying, the more you know, the more there is to know, ya know?!

**Reply by sbwalker** [reply]
@thabaum note that pre-rendering = static rendering (they are the same thing) - which explains why disabling it would cause issues in static render mode ;)

**Reply by sbwalker** [reply]
@thabaum StreamRendering seems to be a "magic pill" in Blazor in .NET 8 :)

"Streaming rendering is a small enhancement over SSR. It allows pages that need to execute long running requests to load quickly via an initial payload of HTML. That HTML can contains placeholders for the areas to be filled with the results of the long running call. Once that long running call completes, the remaining HTML is streamed to the browser and is seamlessly patched into the existing DOM on the client. The cool thing here is that this is all done in the same response. There are no additional calls involved."

However most "magic pills" have side effects... so it will be important to get a better understanding of this feature.

**Reply by thabaum** [reply]
an issue remaining on the backlog is #3829, which updating the page icon in page settings and now found that the same for site name in site settings as well does not seem to reload the UI/page for the user updating these types of settings.   So site settings and page settings updating of page name/logo always updates in the UI without having to hit refresh when clicking save.  If you click on the logo and change it, the site will refresh/reload as needed.  I believe if the site name or page name along with anything relating to appearance is updated the page save behavior should perform the page refresh/reload to ensure a user sees the changes.

It works as expected when you change the logo in site settings from a previous similar commit.  I put up a pull request for making the page edit form save button to just force a reload always on save for pages.  I can work on putting together something different or as you feel fits the bill.   Figured I would point these relating issues out noticeable to the end user experience.  

Take some time to get these new modes down as I can see new possibilities.  Eventually you can look back at all of this and let us all know we are getting close to an RC 5.1.0.  Then we can begin to test out the other parts of the framework that might be getting overlooked right now with module and theme development.  

As with anything we need to use things only as needed.  If it works without adding more or changing default behavior, keep it that way.  I just felt maybe something like this can be changed in the module settings?  A default mode that can be set by a developer as well as the administrator to change according to desired results?  Thinking out loud a little over kill from just adding it to the components that can use one or the the other for desired results in both render modes?  Or a way to set this per component?  Look forward to hearing what you discover for solutions.

One other thing, I was trying it all looking for ways to get rid of the double loading, it is not a fix all and needs to be used with care as it will break the head from rendering.  Some components need it more than others.  In my mind I want as much as possible to render once if it can do the job.  I applied it to the User Profile and made that work normal loading once.  Not all components I notice it much.  There are potentially tradeoffs yet to be discovered.

I know you have a check list to work through, pit crew is on it as needed one day at a time until Oqtane is ready to be off to the races! 🏎️ 

Oqtane may have to come with a warning that it can cause whip lash from the snap of your head back into the seat if you are not fully seated as is it takes off so fast!

**Reply by sbwalker** [reply]
@thabaum the specification of "false" in the attribute [StreamRendering(false)] indicates that StreamRendering should be disabled for the component: "a flag to indicate is this component and its descendants should stream their rendering" - it is not related to pre-rendering (which makes sense). 

So basically by enabling StreamRendering in the Routes and Head components (root components) it means that all components in the entire application are using StreamRendering by default. Based on the earlier challenges encountered with content not rendering properly to the UI, it seems that StreamRendering is required in the vast majority of cases. However, in the case of interactive components (ie. Modules) it appears that disabling StreamRendering can provide a slightly better user experience in some cases, as it eliminates the "double render". 

It would seem to make sense that components which are rendered interactively should disable StreamRendering by default... and components which are rendered statically should enable StreamRendering by default. This could be handled by the framework as part of the UI composition logic, and it could be easily overridden within a specific component if necessary.

Unfortunately, the StreamRendering attribute is a static and can only be specified within the component definition - it cannot be specified on the instance as a property (like rendermode) - which means that you cannot include conditional logic to control whether the attribute is enabled or disabled. The only solution would be to create 2 components - one which includes [StreamRendering(true)] and another which includes  [StreamRendering(false)]... and then you can control which version of the component gets injected. 

I will give this a try this morning to see how to performs.

**Reply by sbwalker** [reply]
@thabaum give PR #3858 a try... it renders interactive module components with StreamRendering disabled and it renders static module components with StreamRendering enabled.

**Reply by sbwalker** [reply]
@thabaum I just posed the question about StreamRendering usage to Microsoft: https://github.com/dotnet/aspnetcore/issues/54157

**Reply by thabaum** [reply]
@sbwalker #3858 is the solution I had thought of as well minus using two components, which created that problem from it being static and unable to set on a component.  Maybe Microsoft can add this ability to set on the component like render modes are as it would make it easier to do what we are doing. 

This works to set this rendering setting and your overall solution using two components to wrap at that part of the pipeline seems to work great!  I can't believe my eyes at how this development ride using Oqtane Framework is getting more options, faster and smoother all at the same time.

I also see you have resolved a number of other issues today and things just keep on rolling forward, awesome, thank you for this exciting journey!

### :rocket:

**Reply by thabaum** [reply]
@sbwalker did I say wow yet?! :)

**Comment by thabaum** [comment]
@sbwalker odd behavior detected while setting themes on different pages, then navigating between the pages, you have to refresh the page each time you click on a navigation.  Almost like if the page theme is different then the site theme a reload or refresh needs to happen.  Only in static SSR.  I suppose I will log the issue.

**Comment by sbwalker** [comment]
There are a number of places within Oqtane where a component needs to redirect to another Url. In Interactive rendering if you placed the call to NavigationManager.NavigateTo() within OnParametersSet() it would throw an exception. So the standard guidance was to include the NavigationManager.NavigateTo() logic in OnAfterRender(). On Static rendering OnAfterRender() is not a life cycle event so it will not be invoked. This means that coditional logic needs to be added in OnParametersSet():

```
if (PageState.RenderMode == RenderModes.Static)
{
    NavigationManager.NavigateTo(url);
}
```

and the traditional NavigationManager.NavigateTo() logic can be retained in OnAfterRender() for interactive scenarios (no conditional logic needed here because it is only called in interactive render mode).

I am not a big fan of having to sprinkle conditional logic around the app based on render mode but at least it is simple to handle in the cases where it is necessary because Oqtane provides the PageState - standard Blazor in .NET 8 does not provide a way to determine the render mode out of the box (it is one of the enhancement requests for .NET 9).

**Reply by sbwalker** [reply]
There are also a number of user experience behaviors in Oqtane which rely on JS Interop. For example, the Login form uses JS Interop to set the focus to the Username field when the form is loaded. These behaviors continue to function correctly because the majority of modules are using interactive rendering. 

However if you created a module which needed to work in static rendering, you could not rely on JS Interop as it can only be called from OnAfterRender(). So you would need to create an alternative method. I assume that in static rendering you can just embed standard JavaScript in your components - which makes it simpler. There is nothing in Oqtane which requires this at the moment but I am sure it will come up in the near future.

**Comment by sbwalker** [comment]
Even with the improvements outlined at the top of this thread, the run-time characteristics of static rendering are much different than interactive rendering. This is because static rendering is stateless - which means methods such as SiteService GetSite() need to be called on every browser request to load the data from the server (whereas interactive rendering is stateful and can re-use data loaded previously).

Caching had been implemented in prior releases within the SiteService GetSite() method for unauthenticated users which eliminated the bulk of database calls for public-facing scenarios. This was possible because all unauthenticated users have the same set of permissions in terms of the content they are allowed to access. However, this does not provide any benefit for authenticated users.

Caching data for each authenticated user is not difficult... the challenge is dealing with cache invalidation when the site configuration changes at run-time, or when the user's permissions are updated. This required some improvements to the caching infrastructure to provide more context and control over cache items. With the latest changes, authenticated users will get the same performance benefits as unauthenticated users, and the cache will be refreshed immediately when information changes.
